packages/eslint-plugin/src/rules/ban-imports.js

```diff
@@ -88,61 +90,56 @@ module.exports = createRule({
-        importOrExport.source.type !== AST_NODE_TYPES.Literal ||
-        typeof importOrExport.source.value !== 'string'
-      ) {
-        return;
-      }
-
-      const importPath = importOrExport.source.value;
-      const specifiers =
-        importOrExport.type === AST_NODE_TYPES.ExportNamedDeclaration
-          ? importOrExport.specifiers
-          : // Filter out default imports and namespace (star) imports
-            /** @type {ImportSpecifier[]} */ (importOrExport.specifiers.filter(
-              spec => spec.type === AST_NODE_TYPES.ImportSpecifier,
-            ));
+    function isMatch(value, stringOrRegex) {
+      return typeof stringOrRegex === 'string' ? value === stringOrRegex : stringOrRegex.test(value);
+    }
 
+    /**
+     * @param {ImportDeclaration | ExportNamedDeclaration} importOrExport the whole import/export node
+     * @param {string} importPath path importing/exporting from
+     * @param {Identifier[]} identifiers imported/exported identifiers
+     */
+    function checkImportOrExport(importOrExport, importPath, identifiers) {
       for (const rule of options) {
-        const { path, pathRegex, names, message = '' } = rule;
-        const pathForErrors = path || /** @type {RegExp} */ (pathRegex).source;
-        const messageForErrors = message && `: ${message}`;
+        const { path, names, message } = rule;
+
+        if (!isMatch(importPath, path)) {
+          continue;
+        }
+
+        const errorData = {
+          path: typeof path === 'string' ? path : path.source,
+          message: message ? `: ${message}` : '',
+          verb: importOrExport.type === AST_NODE_TYPES.ImportDeclaration ? 'Importing' : 'Exporting',
+        };
 
-        if ((path && path === importPath) || (pathRegex && pathRegex.test(importPath))) {
-          if (!names) {
-            // All imports from this path are banned
-            context.report({
-              node: importOrExport,
-              messageId: 'pathNotAllowed',
-              data: { path: pathForErrors, message: messageForErrors },
-            });
-          } else {
-            // Only certain imports from this path are banned
-            for (const spec of specifiers) {
-              const identifier = spec.type === AST_NODE_TYPES.ExportSpecifier ? spec.local : spec.imported;
-              if (
-                names.some(name => (typeof name === 'string' ? name === identifier.name : name.test(identifier.name)))
-              ) {
-                context.report({
-                  node: identifier,
-                  messageId: 'nameNotAllowed',
-                  data: { path: pathForErrors, name: identifier.name, message: messageForErrors },
-                });
-              }
+        if (!names) {
+          // All imports from this path are banned
+          context.report({
+            node: importOrExport,
+            messageId: 'pathNotAllowed',
+            data: errorData,
+          });
```

Comment:
I'm not sure early exiting would help clarity in this case, though I do think it would make sense to get rid of the negation and reverse the order. Whether functional forEach or imperative iteration is more readable is somewhat a matter of opinion, or may vary case-by-case.